### PR TITLE
Fix optional list

### PR DESCRIPTION
**Pull request by Dmitrii-Shevchenko**
Added ForceSeverity as ReadError.
Changed method hasErrors.

### CODE REVIEW

File: core/src/main/scala/zio/config/ReadError.scala [hunk 1]
```diff
@@ -3,20 +3,22 @@ package zio.config
...
-    case ReadError.FormatError(path, message)     => s"FormatError(${path}, ${message})"
-    case ReadError.ConversionError(path, message) => s"ConversionError(${path},${message}"
-    case ReadError.OrErrors(list)                 => s"OrErrors(${list.map(_.toString)})"
-    case ReadError.AndErrors(list)                => s"AndErrors(${list.map(_.toString)}"
+    case ReadError.MissingValue(path)                   => s"MissingValue(${path})"
+    case ReadError.FormatError(path, message)           => s"FormatError(${path}, ${message})"
+    case ReadError.ConversionError(path, message)       => s"ConversionError(${path},${message}"
+    case ReadError.OrErrors(list)                       => s"OrErrors(${list.map(_.toString)})"
+    case ReadError.AndErrors(list)                      => s"AndErrors(${list.map(_.toString)}"
+    case ReadError.ForceSeverity(error, treatAsMissing) => s"AndErrors(${error.toString},${treatAsMissing})"
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadError.scala]: invalid name

File: core/src/main/scala/zio/config/ReadError.scala [hunk 2]
```diff
@@ -35,6 +39,7 @@ sealed trait ReadError[A] extends Exception { self =>
       case ReadError.ConversionError(_, _) => 1
       case ReadError.OrErrors(list)        => list.map(_.size).sum
       case ReadError.AndErrors(list)       => list.map(_.size).sum
+      case ReadError.ForceSeverity(_, _)   => 1
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadError.scala]: Use `size` of inner error

File: core/src/main/scala/zio/config/ReadError.scala [hunk 3]
```diff
@@ -100,15 +105,24 @@ sealed trait ReadError[A] extends Exception { self =>
...
+          )
+        )
+      )
+
     def readErrorToSequential[A](readError: ReadError[A]): Sequential =
       readError match {
         case r: ReadError.MissingValue[A]    => renderMissingValue(r)
         case r: ReadError.FormatError[A]     => renderFormatError(r)
         case r: ReadError.ConversionError[A] => renderConversionError(r)
+        case r: ReadError.ForceSeverity[A]   => renderForceSeverity(r)
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadError.scala]: Just call `readErrorToSequential` on inner error.

File: core/src/main/scala/zio/config/ReadModule.scala [hunk 4]
```diff
@@ -21,15 +21,15 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
 
     def loopDefault[B](path: List[Step[K]], keys: List[K], cfg: Default[B]): Res[B] =
       loopAny(path, keys, cfg.config) match {
-        case Left(error) if hasParseErrors(error) || hasConversionErrors(error) =>
+        case Left(error) if hasErrors(error) =>
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadModule.scala]: Rename to `hasUnrecoverableErrors`

File: core/src/main/scala/zio/config/ReadModule.scala [hunk 5]
```diff
@@ -164,17 +164,12 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
...
-      .lift(value)
-      .getOrElse(false)
-
-  private def hasParseErrors[K](error: ReadError[K]): Boolean =
-    hasErrors(error)({ case ReadError.FormatError(_, _) => true })
-
-  private def hasConversionErrors[K](error: ReadError[K]): Boolean =
-    hasErrors(error)({ case ReadError.ConversionError(_, _) => true })
+  ): Boolean = value match {
+    case ReadError.ForceSeverity(_, _)   => true
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadModule.scala]: Use `treatAsMissing` here.

File: core/src/main/scala/zio/config/ReadModule.scala [hunk 6]
```diff
@@ -164,17 +164,12 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
...
+    case ReadError.FormatError(_, _)     => true
+    case ReadError.ConversionError(_, _) => true
+    case OrErrors(_)                     => true
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadModule.scala]: Check inner errors.

File: core/src/main/scala/zio/config/ReadModule.scala [hunk 7]
```diff
@@ -164,17 +164,12 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
...
+    case AndErrors(_)                    => true
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadModule.scala]: Check inner errors.

**Review by senia-psm — CHANGES_REQUESTED** [review]
Fix `loopMap`.
Add tests for `map`.
Run all tests (`sbt testAll`) and fix failed tests.

File: core/src/main/scala/zio/config/ReadError.scala (hunk 1 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadError.scala]: Fixed

File: core/src/main/scala/zio/config/ReadError.scala (hunk 2 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadError.scala]: Fixed

File: core/src/main/scala/zio/config/ReadError.scala (hunk 3 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadError.scala]: Fixed

File: core/src/main/scala/zio/config/ReadModule.scala (hunk 4 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadModule.scala]: Fixed

File: core/src/main/scala/zio/config/ReadModule.scala (hunk 5 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadModule.scala]: Fixed

File: core/src/main/scala/zio/config/ReadModule.scala (hunk 6 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadModule.scala]: Fixed

File: core/src/main/scala/zio/config/ReadModule.scala (hunk 7 above)
Comment (Dmitrii-Shevchenko) [reply core/src/main/scala/zio/config/ReadModule.scala]: Fixed

File: core/src/main/scala/zio/config/ReadError.scala [hunk 8]
```diff
@@ -105,10 +110,10 @@ sealed trait ReadError[A] extends Exception { self =>
         case r: ReadError.MissingValue[A]    => renderMissingValue(r)
         case r: ReadError.FormatError[A]     => renderFormatError(r)
         case r: ReadError.ConversionError[A] => renderConversionError(r)
+        case r: ReadError.ForceSeverity[A]   => readErrorToSequential(r)
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadError.scala]: Infinite recursion.

File: core/src/test/scala/zio/config/ListAndOptionalTest.scala [hunk 9]
```diff
@@ -96,6 +96,22 @@ object ListAndOptionalTest
...
         },
+        test("key doesn't exist in list") {
+          val src = ConfigSource.fromPropertyTree(
+            PropertyTree.Sequence(List(Record(Map()))),
+            "src"
+          )
+          val optional: ConfigDescriptor[Option[List[String]]] = list(string("keyNotExists")).optional
+          assert(read(optional from src))(isLeft(anything))
+        },
+        test("key doesn't exist in map") {
```
Comment (senia-psm) [inline core/src/test/scala/zio/config/ListAndOptionalTest.scala]: move this test to another file - this file is only for `list`, not for `map`

**Pull request merged**